Custom bindings doc + iou-no-codegen scala example (#1569)
digital-asset/daml: docs/source/app-dev/bindings-x-lang/index.rst

.. Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
.. SPDX-License-Identifier: Apache-2.0

Creating your own bindings
##########################

This page gets you started with creating custom bindings for the Digital Asset distributed ledger.

Introduction
============

Digital Asset currently provides bindings for the following programming languages:

- :doc:`Java </app-dev/bindings-java/index>`

- :doc:`Scala </app-dev/bindings-scala/index>`

- :doc:`JavaScript (Node.js) </app-dev/bindings-js>`

You can create bindings for any programming language supported by `gRPC <https://grpc.io/docs/>`_.

What do we mean by "bindings"? Bindings for a language consist of two main components:

- Ledger API

>     Client "stubs" for the programming language, -- the remote API that allows sending ledger commands and receiving ledger transactions. You have to generate **Ledger API** from `the gRPC protobuf definitions in the daml repository on GitHub <https://github.com/digital-asset/daml/tree/master/ledger-api/grpc-definitions>`_. **Ledger API** is documented on this page: :doc:`/app-dev/grpc/index`. The `gRPC <https://grpc.io/docs/>`_ tutorial explains how to generate client "stubs".

Reviewer: Big improvement, this is much clearer!

Note that for different reasons we chose codegen, but that is not the only option. There is really a broad category of metaprogramming features that can solve this problem just as well or even better than codegen; they are language-specific, but often much easier to maintain (i.e. no need to add a build step). Some examples are:

- `F# Type Providers <https://docs.microsoft.com/en-us/dotnet/fsharp/tutorials/type-providers/creating-a-type-provider#a-type-provider-that-is-backed-by-local-data>`_

- `Template Haskell <https://wiki.haskell.org/Template_Haskell>`_

- Scala macro annotations (not future-proof enough to use when implementing the last Scala codegen)

Building Ledger Commands
========================

> No matter what approach you take, either manually building commands or writing a codegen to do this, you need to understand how ledger commands are structured. This section demonstrates how to build create and exercise commands manually and how it can be done using contract classes generated by Scala codegen.

Reviewer: Also a massive improvement. Nice!

- Codegen
    Code generator. A program that generates classes representing DAML contract templates in X language. These classes incorporate all boilerplate code responsible for constructing: (:ref:`com.digitalasset.ledger.api.v1.CreateCommand`) and (:ref:`com.digitalasset.ledger.api.v1.ExerciseCommand`) corresponding for each DAML contract template.

Technically codegen is optional. It is possible to construct (:ref:`com.digitalasset.ledger.api.v1.CreateCommand`) and (:ref:`com.digitalasset.ledger.api.v1.ExerciseCommand`) manually. However this is very tedious and error-prone.

Here is an example of constructing a :ref:`com.digitalasset.ledger.api.v1.CreateCommand` for contract template with one field: `party` ::

  val fields = Seq(RecordField("party", Some(Value(party))))
  val command =
    Command.Create(
      CreateCommand(
        templateId = Some(templateId),
        createArguments = Some(Record(Some(templateId), fields))))

> if you do not specify `party` field or type it's name or value incorrectly, the above code will compile but fail in run-time because you did not structure you create command correctly. Codegen is expected to simplify the command construction by providing an auto-generated utilities to help you constructing commands. E.g. here is how a similar contract instantiation would look if you are using a scala codegen to generate contract classes::

Reviewer 2: ```suggestion
If you do not specify `party` field or type its name or value incorrectly, or don't order the fields exactly as they are in the DAML template, the above code will compile but fail at run-time because you did not structure your create command correctly. Codegen is expected to simplify the command construction by providing auto-generated utilities to help you construct commands. Here is how a similar contract instantiation would look if you are using Scala codegen to generate contract classes::
```

.. Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
.. SPDX-License-Identifier: Apache-2.0

Custom bindings
###############

This page gets you started with creating custom bindings for the Digital Asset distributed ledger.

Introduction
============

Digital Asset currently provides bindings for the following programming languages:

- :doc:`Java </app-dev/bindings-java/index>`

- :doc:`Scala </app-dev/bindings-scala/index>`

- :doc:`Node.js JavaScript </app-dev/bindings-js>`

However it is possible to create bindings for any programming language supported by `gRPC <https://grpc.io/docs/>`_.

What exactly is "bindings"? X language bindings consists of two main components:

- Ledger API

>     Client "stubs" for X language, -- the remote ledger API that allows sending ledger commands and receiving ledger transactions. Generated from gRPC protobuf definitions in the **daml** repository on GitHub: https://github.com/digital-asset/daml/tree/master/ledger-api/grpc-definitions and documented on this page: :doc:`The Ledger API using gRPC </app-dev/grpc/index>`

Author: going to accept parts of the suggested changes. I want explain that client stubs, remote API and Ledger API are the same thing

>     Client "stubs" for X language, -- the remote ledger API that allows sending ledger commands and receiving ledger transactions. Generated from gRPC protobuf definitions in the **daml** repository on GitHub: https://github.com/digital-asset/daml/tree/master/ledger-api/grpc-definitions and documented on this page: :doc:`The Ledger API using gRPC </app-dev/grpc/index>`

Reviewer: Great, that sounds really helpful

- Codegen
    Code generator. A program that generates classes representing DAML contract templates in X language. These classes incorporate all boilerplate code for constructing: :ref:`com.digitalasset.ledger.api.v1.CreateCommand` and :ref:`com.digitalasset.ledger.api.v1.ExerciseCommand` corresponding for each DAML contract template.

> Technically codegen is optional. It is possible to construct the commands manually. However it is very tedious and error-prone.

Author: I don't agree with this... Codegen will be an overkill for a small project with 5-10 templates. You will spend 2-3 months creating a proper codegen vs 2-3 weeks of manually constructing the commands. It all depends on what users want to achieve.

> Technically codegen is optional. It is possible to construct the commands manually. However it is very tedious and error-prone.

Reviewer: Ah, I see - I over-extrapolated from what you wrote. In that case could you edit it so it reflects the comment you just made? Maybe something like it makes sense for a big/long-term project, but for short projects is overkill

> Technically codegen is optional. It is possible to construct the commands manually. However it is very tedious and error-prone.

Author: yep... done. Let me know if it read ok.

> Building Create Command

Author: updated

=======================

Let's recall an **IOU** example from the :doc:`Quickstart guide </getting-started/quickstart>`, where `Iou` template is defined like this:

.. literalinclude:: ./code-snippets/quickstart/template-root/daml/Iou.daml
  :language: daml
  :lines: 9-15

> Here is how to manually build :ref:`com.digitalasset.ledger.api.v1.CreateCommand` for the above contract template in Scala:

Reviewer: Tiny grammar tweak
```suggestion
Here is how to manually build a :ref:`com.digitalasset.ledger.api.v1.CreateCommand` for the above contract template in Scala:
```

.. literalinclude:: ./code-snippets/iou-no-codegen/application/src/main/scala/com/digitalasset/quickstart/iou/IouCommands.scala
   :start-after: // <doc-ref:iou-no-codegen-create-command>
   :end-before: // </doc-ref:iou-no-codegen-create-command>

> If you do not specify any of the above fields or type their names or values incorrectly, or do not order them exactly as they are in the DAML template, the above code will compile but fail at run-time because you did not structure your create command correctly. Codegen is expected to simplify the command construction by providing auto-generated utilities to help you construct commands.

Reviewer: Just suggesting breaking this up slightly differently. 
```suggestion
If you do not specify any of the above fields or type their names or values incorrectly, or do not order them exactly as they are in the DAML template, the above code will compile but fail at run-time because you did not structure your create command correctly.

Codegen will usually simplify the command construction by providing auto-generated utilities to help you construct commands. For example, when you use :doc:`Scala codegen </app-dev/bindings-scala/index>` to generate contract classes, a similar contract instantiation would look like this:
```

> When you use :doc:`Scala codegen </app-dev/bindings-scala/index>` to generate contract classes, a similar contract instantiation would look:

Reviewer: Removing as I've added it above.
```suggestion
```

   :start-after: // <doc-ref:iou-contract-instance>
   :end-before: // </doc-ref:iou-contract-instance>

Building Exercise Command
=========================

To build :ref:`com.digitalasset.ledger.api.v1.ExerciseCommand` for `Iou_Transfer`:

.. literalinclude:: ./code-snippets/quickstart/template-root/daml/Iou.daml
  :language: daml
  :lines: 23, 52-55

manually in Scala:

.. literalinclude:: ./code-snippets/iou-no-codegen/application/src/main/scala/com/digitalasset/quickstart/iou/IouCommands.scala
   :start-after: // <doc-ref:iou-no-codegen-exercise-command>
   :end-before: // </doc-ref:iou-no-codegen-exercise-command>

versus creating the same command using a value class generated by :doc:`Scala codegen </app-dev/bindings-scala/index>`:

.. literalinclude:: ./code-snippets/quickstart-scala/application/src/main/scala/com/digitalasset/quickstart/iou/IouMain.scala
   :start-after: // <doc-ref:iou-exercise-transfer-cmd>
   :end-before: // </doc-ref:iou-exercise-transfer-cmd>

Summary
=======

If you are creating custom bindings for the Digital Asset distributed ledger, you will need to:

> - generate **Ledger API** from the gRPC definitions

Reviewer: This is a little bit light on detail. Is there anything we can link to in the gRPC docs to explain how to do this? Otherwise it's a bit ["draw the rest of the owl"](https://www.google.com/search?q=draw+the+owl&source=lnms&tbm=isch&sa=X&ved=0ahUKEwiC5ZDviOHiAhXTwsQBHflmAE0Q_AUIECgB&biw=1280&bih=721#imgrc=gbPGHaXIjOhvjM:).

Author: it is just a summary to re-iterate what needs to be done. The **Ledger API** doc and the link to gRPC tutorial above cover this up.

Author: I am also fine with removing this completely.

> - decide if you write a codegen to generate contract classes or you manually build commands for all contracts defined in your DAML model.

Reviewer 2: ```suggestion
- decide whether to write a codegen to generate contract classes or manually build commands for all contracts defined in your DAML model.
```

> The above examples should help you get started. Feel free to contact us on the DAML community `Slack <https://damldriven.slack.com/sso/saml/start>`_ or open a `GitHub issue <https://github.com/digital-asset/daml/issues>`_.

Reviewer: Totally up to you, but it might be better to link to the support page instead (`/support/support`). That way, we have one centralised way of telling people how to ask for help.

- :doc:`Java </app-dev/bindings-java/index>`

- :doc:`Scala </app-dev/bindings-scala/index>`

- :doc:`JavaScript (Node.js) </app-dev/bindings-js>`

However, you can create bindings for any programming language supported by `gRPC <https://grpc.io/docs/>`_.

What do we mean by "bindings"? Bindings for a language consist of two main components:

- Ledger API
    Client "stubs" for the programming language, -- the remote API that allows sending ledger commands and receiving ledger transactions. You have to generate **Ledger API** from `the gRPC protobuf definitions in the daml repository on GitHub <https://github.com/digital-asset/daml/tree/master/ledger-api/grpc-definitions>`_. **Ledger API** is documented on this page: :doc:`/app-dev/grpc/index`. The `gRPC <https://grpc.io/docs/>`_ tutorial explains how to generate client "stubs".

- Codegen
     A code generator is a program that generates classes representing DAML contract templates in the language. These classes incorporate all boilerplate code for constructing: :ref:`com.digitalasset.ledger.api.v1.CreateCommand` and :ref:`com.digitalasset.ledger.api.v1.ExerciseCommand` corresponding for each DAML contract template.

> Technically codegen is optional. You can construct the commands manually from the auto-generated **Ledger API** classes. However it is very tedious and error-prone. If you are creating ah hoc bindings for a project with a few contract templates, writing a proper codegen may be an overkill. On the other hand if you have hundreds of contract templates in your project or planning to build language bindings that you will share across multiple projects, we recommend including codegen in your bindings. It will save you time in the long run.

Reviewer 2: It's a little more difficult to talk about, but there's really a broad category of metaprogramming features that can solve this problem just as well or even better than codegen; they're language-specific, but often much easier to maintain (i.e. no need to add a build step). Some examples are

1. F# [type providers](https://docs.microsoft.com/en-us/dotnet/fsharp/tutorials/type-providers/creating-a-type-provider#a-type-provider-that-is-backed-by-local-data),
2. Template Haskell,
3. Scala macro annotations (not future-proof enough to use when implementing the last Scala codegen),

or pretty much any sort of metaprogramming mechanism that works opposite to the usual unsafe "reflection", working to _produce_ types instead of _inspect_ them, at _compile_ time instead of runtime.

I don't know how much we want to say about that, though.

Author: @S11001001 are you suggesting we remove this entire section of just leave it like:
> Technically codegen is optional. You can construct the commands manually from the auto-generated **Ledger API** classes. However it is very tedious and error-prone

Author: but I think we need to give readers a hint that you can solve the same problem with either codegen, macros or metaprogramming if you don't want to write your commands manually.

Author: @S11001001, how about this explantion for the `codgen`. A code generator is a program or any metaprogramming technic that generates ledger commands  at compile-time.

Reviewer 2: > are you suggesting we remove this entire section

Not at all, you're right ⬇️ 

> we need to give readers a hint that you can solve the same problem with either codegen, macros or metaprogramming if you don't want to write your commands manually

Because nowhere else have we explained why we bother writing codegens, it's just something _we_ know.

Also, notwithstanding a rewrite to talk about other metaprogramming forms:

```suggestion
Technically codegen is optional. You can construct the commands manually from the auto-generated **Ledger API** classes. However, it is very tedious and error-prone. If you are creating *ad hoc* bindings for a project with a few contract templates, writing a proper codegen may be overkill. On the other hand, if you have hundreds of contract templates in your project or are planning to build language bindings that you will share across multiple projects, we recommend including a codegen in your bindings. It will save you and your users time in the long run.
```

Reviewer 2: > how about this explantion for the `codgen`. A code generator is a program or any metaprogramming technic that generates ledger commands at compile-time.

No, that would be very misleading. I mentioned the other metaprogramming techniques because they _aren't_ codegens.

Author: added codegen alternatives
